nemo/collections/asr/parts/utils/nmesc_clustering.py

```diff
@@ -27,75 +27,312 @@
+        distance = distance_diff.pow(2).sum(dim=1).view(N_ci, -1)
+        distance_to_candidates = torch.minimum(closest_dist_sq, distance)
+        candidates_pot = distance_to_candidates.sum(dim=1)
 
-try:
-    from torch.linalg import eigh as eigh
+        best_candidate = torch.argmin(candidates_pot)
+        current_pot = candidates_pot[best_candidate]
+        closest_dist_sq = distance_to_candidates[best_candidate]
+        best_candidate = candidate_ids[best_candidate]
 
-    TORCH_EIGN = True
+        centers[c] = X[best_candidate]
+        indices[c] = best_candidate
+    return centers, indices
 
-except ImportError:
-    TORCH_EIGN = False
-    from scipy.linalg import eigh as eigh
 
-    logging.warning("Using eigen decomposition from scipy, upgrade torch to 1.9 or higher for faster clustering")
+@torch.jit.script
+def kmeans_torch(
+    X: torch.Tensor,
+    num_clusters: int,
+    threshold: float = 1e-4,
+    iter_limit: int = 15,
+    device: torch.device = torch.device('cpu'),
+):
+    """
+    Run k-means algorithm on the given set of spectral embeddings in X. The threshold
+    and iter_limit variables are set to show the best performance on speaker diarization
+    tasks. The overall implementation of k-means algorithm is inspired by the k-means
+    algorithm implemented in https://github.com/scikit-learn/scikit-learn.
+
+    References:
+        Arthur, David, and Sergei Vassilvitskii. k-means++: The advantages of careful
+        seeding. Proceedings of the eighteenth annual ACM-SIAM symposium on Discrete
+        algorithms, Society for Industrial and Applied Mathematics (2007).
+
+    Args:
+        X: (torch.tensor)
+            Cosine similarity matrix calculated from speaker embeddings
 
+        num_clusters: (int)
+            The estimated number of speakers.
 
-def isGraphFullyConnected(affinity_mat):
-    return getTheLargestComponent(affinity_mat, 0).sum() == affinity_mat.shape[0]
+        threshold: (float)
+            This threshold limits the change of center values. If the square of
+            the center shift values are bigger than this threshold, the iteration stops.
 
+        iter_limit: (int)
+            The maximum number of iterations that is allowed by the k-means algorithm.
 
-def getTheLargestComponent(affinity_mat, seg_index):
+        device: (torch.device)
+            Torch device variable
+
+    Returns:
+        selected_cluster_indices: (torch.tensor)
+            The assigned cluster labels from the k-means clustering.
+    """
+    # Convert tensor type to float
+    X = X.float().to(device)
+    input_size = X.shape[0]
+
+    # Initialize the cluster centers with kmeans_plusplus algorithm.
+    plusplus_init_states = kmeans_plusplus_torch(X, n_clusters=num_clusters, random_state=0, device=device)
+    centers = plusplus_init_states[0]
+
+    iter_count = 0
+    selected_cluster_indices = torch.zeros(input_size).int()
```

Comment:
I don't get it. why converting above to float and use int here?

Reply:
selected_cluster_indices is cluster index as the name suggests. So it needs to be int. While spectral embedding X is converted to float() to match the type with other operations.